### ISSUE TITLE
Auto Apply Coupon | Ignore HTTP_REFERER check if it starts with HTTPS_SERVER

### ISSUE THREAD

**Issue by asiksami** [zencart/zencart#6193]
This is a nice addon, thanks @neekfenwick. In the future, this could work nicely on QR-codes / newsletters links.

Thought I didn't find a way to disable it in any way. It might not take that much resources on checking `$_GET['coupon_code']`, but if One is sure that it will not be used in its store or someone starts to exploit it somehow, you would like to have it disabled somewhere that it doesn't require code changes.

Would it be good to have this coupon auto apply to be disabled somewhere? Maybe soft configure to site-specific-overrides.php? Or maybe better as a database configure variable?

**Comment by neekfenwick** [comment]
Hi @asiksami, I'd have to ask the likes of @drbyte about whether we should have a global kill-switch for the coupon auto apply feature and how best to implement it.  As you say, a `site_specific_overrides.php` flag or a new `configure` table database entry.  Because `ot_coupon` module may not be 'installed' i.e. the database entries it creates may not exist, then a `site_specific_overrides.php` flag seems wiser.

    $ot_coupon_coupon_code_auto_apply_disabled = true;

Bear in mind that two values are used in `init_coupons.php` which could theoretically be abused: `$_GET['coupon_code']`, and `$_SERVER['HTTP_REFERER']`, both of which are sanitized before use.

I am reluctant to implement further defenses such as this kill-switch, it seems over-cautious, and it's no more exploitable than any other query parameter that may be passed in.

What do you think @drbyte ?

I guess we don't need to consider specific sanitizing in `init_sanitize.php` for the `coupon_code` query param?  If it's provided, then the new code in `ini_coupons.php` will try to look it up, and it does use `zen_db_prepare_input` to sanitize it there.

**Comment by scottcwilson** [comment]
> I am reluctant to implement further defenses such as this kill-switch, it seems over-cautious, and it's no more exploitable than any other query parameter that may be passed in.

I agree.  I think it's unlikely that this specific feature would be a target of attackers.  Most of what I see is URLs with embedded encoded queries, much of which is silently blocked by the latest releases.

**Comment by neekfenwick** [comment]
So @asiksami it seems this is a case to close as WontFix, although one thought comes to me.

`init_coupons.php` currently queries the `coupons.referrer` field with a `LIKE` match for every single page request, but most page requests will be internal links between a site's own pages.  I just tested a click on my dev site and the `HTTP_REFERER` header was `https://zc158.orig/` in a request from the home page, for example.  So, the init can ignore the referrer check for any value that begins with `HTTPS_SERVER`.

We could re-purpose this Issue for this :) What do you think?

**Comment by asiksami** [comment]
Sure @neekfenwick. Renamed the issue.

I think too that these variables are not that much exploited. But when thinking this as performance wise, when these checks are done on every call, on page loads, on ajax call (?) etc, it might start to slow things down on high traffic sites.

This `HTTP_REFERER` check for `HTTPS_SERVER` would definitely help to skip unnecessary queries to `coupons`. Also should keep mind that `coupons` table can get big on some cases. Ie. if one uses customer based coupons where you create one coupon for each customer.

**Comment by drbyte** [comment]
I agree that this fix certainly helps limit bottlenecks caused by "self".

@neekfenwick perhaps it is indeed worth exploring a different db structure for this after all?  
ie: a table of domain names with corresponding coupon ids. One record per domain, with a `unique` index. 
That way a LIKE query is avoided, and db indexes can be used.

The lookup query can be a join on the coupon_id to get the coupon_code.

The admin page where the CSV data is collected would have to sanitize and trim the domains before storing the records, and of course store the multiple records instead of simply adding to the (now-current but not officially released) referrers field in the coupons table. 

Aside: That would also make using external tools for importing such data slightly easier to implement.

**Comment by neekfenwick** [comment]
> @neekfenwick perhaps it is indeed worth exploring a different db structure for this after all?
> ie: a table of domain names with corresponding coupon ids. One record per domain, with a `unique` index.

Imagining the UI required for a list of domains; supporting saving an array i.e. one-to-many table relationship can be quite a lot of work, especially on the UI. I've used various data binding libraries in the past but I don't think ZC uses one as standard, right? I know we're not doing anything as advanced as virtual DOM, JSX etc, but is there a 'add element'/'remove element' you'd recommend or do I have to write one from scratch?

**Comment by drbyte** [comment]
> > @neekfenwick perhaps it is indeed worth exploring a different db structure for this after all?
> > ie: a table of domain names with corresponding coupon ids. One record per domain, with a `unique` index.
> 
> Imagining the UI required for a list of domains; supporting saving an array i.e. one-to-many table relationship can be quite a lot of work, especially on the UI. I've used various data binding libraries in the past but I don't think ZC uses one as standard, right? I know we're not doing anything as advanced as virtual DOM, JSX etc, but is there a 'add element'/'remove element' you'd recommend or do I have to write one from scratch?

Good point. Collecting data for "creating" those records is simple (even just let them provide comma-separated text as you do now, and then split/parse/save). But edit/update/delete in some sort of intuitive UI is a zany amount of work for the few times people will touch this. But, I suppose you could clone the admin Countries or Currencies pages for coupon-referrers mgmt. First column would be the referrer domain, and second column a drop-down to select from available coupons.

While the few stores who use this will love a nice UI for it, the majority of stores will never use it (even though many "should" use marketing approaches such as this).

**Comment by neekfenwick** [comment]
> First column would be the referrer domain, and second column a drop-down to select from available coupons.

Oh, you meant one-to-many from domain to coupons?  I assumed the other way around, similar to the way coupon Restrictions work; a coupon has a one-to-many relationship with products or categories.  So, a coupon would have many referring domains.  If two coupons need the same domains, we just duplicate the referrer data.

To simplify the edit/update/delete you mention, the coupon Admin UI could have a list of textboxes for Referrers, instead of just one, and the POST handler deal with data updates by deleting all referrers for the coupon and re-inserting whatever was POSTed.

I'm aware that jQuery has a `.widget()` system for declaring reusable widgets, it may make sense to have our own widget for an array of inputs with the +/- button handling.  I imagine we'd need to create a new folder `admin/javascript/widgets` for the .js files involved, but I'm not experienced with jQuery file structure and packaging being more of an AMD/Dojo developer.

**Comment by drbyte** [comment]
> To simplify the edit/update/delete you mention, the coupon Admin UI could have a list of textboxes for Referrers, instead of just one, and the POST handler deal with data updates by deleting all referrers for the coupon and re-inserting whatever was POSTed.

Ya, that kind of UI could work directly within the coupon-edit page. 
A little bit hard to find a referrer if you wanted to edit it though.

**Comment by drbyte** [comment]
> If two coupons need the same domains, we just duplicate the referrer data.

The current implementation only searches for the first match of a referrer domain. Not sure what criteria you'd use to assign/apply the coupon for a subsequent match.

**Comment by scottcwilson** [comment]
Seems like this would have to be unique for enabled coupons.

**Comment by neekfenwick** [comment]
> > If two coupons need the same domains, we just duplicate the referrer data.
> 
> The current implementation only searches for the first match of a referrer domain. Not sure what criteria you'd use to assign/apply the coupon for a subsequent match.

Are you worried that a ZC site may have multiple coupons for the same referrer?  I realise now, that's a problem.  i.e. In the situation of loading a page, needing to check if the current `HTTP_REFERER` value of `abc.com` matches any coupons, the site has configured COUPON-A and COUPON-B both with the referrer 'abc.com', technically you can only have one coupon applied at a time (in `$_SESSION['cc_id']`) and so the system doesn't know which of the two coupons to apply.  Seems to me that we should make it an error to try to save a coupon with a referrer already configured for another coupon.  Is that OK?

**Comment by drbyte** [comment]
Yes. I think an error message makes sense. Perhaps state what coupon code is already assigned to it, for easier correction.

**Comment by neekfenwick** [comment]
> Ya, that kind of UI could work directly within the coupon-edit page. A little bit hard to find a referrer if you wanted to edit it though.

Yeah, there's no search mechanism proposed to find a coupon in coupon-admin for a specific referrer.

Please see PR #6211 with a couple of improvements covering what we've discussed here.  Feedback etc welcome :)
